src/librustc_incremental/persist/fs.rs

```diff
@@ -0,0 +1,862 @@
+                         -> Option<PathBuf> {
+    let iter = crate_dir.read_dir()
+                        .unwrap() // FIXME
+                        .filter_map(|e| e.ok().map(|e| e.path()));
+
+    find_source_directory_in_iter(iter, source_directories_already_tried)
+}
+
+fn find_source_directory_in_iter<I>(iter: I,
+                                    source_directories_already_tried: &FnvHashSet<PathBuf>)
+                                    -> Option<PathBuf>
+    where I: Iterator<Item=PathBuf>
+{
+    let mut best_candidate = (UNIX_EPOCH, None);
+
+    for session_dir in iter {
+        if source_directories_already_tried.contains(&session_dir) ||
+           session_dir.to_string_lossy().ends_with("-working") {
+            continue
+        }
+
+        let timestamp = {
+            let directory_name = session_dir.file_name().unwrap().to_string_lossy();
+            assert_no_characters_lost(&directory_name);
+
+            extract_timestamp_from_session_dir(&directory_name)
+                .unwrap_or_else(|_| {
+                    bug!("unexpected incr-comp session dir: {}", session_dir.display())
+                })
+        };
+
+        if timestamp > best_candidate.0 {
+            best_candidate = (timestamp, Some(session_dir));
+        }
+    }
+
+    best_candidate.1
+}
+
+fn extract_timestamp_from_session_dir(directory_name: &str)
+                                      -> Result<SystemTime, ()> {
+    if !directory_name.starts_with("sess-") {
+        return Err(())
+    }
+
+    let dash_indices: Vec<_> = directory_name.match_indices("-")
+                                             .map(|(idx, _)| idx)
+                                             .collect();
+    if dash_indices.len() < 2 {
+        return Err(())
+    }
+
+    string_to_timestamp(&directory_name[dash_indices[0]+1 .. dash_indices[1]])
+}
+
+fn timestamp_to_string(timestamp: SystemTime) -> String {
+    let duration = timestamp.duration_since(UNIX_EPOCH).unwrap();
+    let nanos = duration.as_secs() * 1_000_000_000 +
+                (duration.subsec_nanos() as u64);
+    format!("{:x}", nanos)
+}
+
+fn string_to_timestamp(s: &str) -> Result<SystemTime, ()> {
+    let nanos_since_unix_epoch = u64::from_str_radix(s, 16);
+
+    if nanos_since_unix_epoch.is_err() {
+        return Err(())
+    }
+
+    let nanos_since_unix_epoch = nanos_since_unix_epoch.unwrap();
+
+    let duration = Duration::new(nanos_since_unix_epoch / 1_000_000_000,
+                                 (nanos_since_unix_epoch % 1_000_000_000) as u32);
+    Ok(UNIX_EPOCH + duration)
+}
+
+fn crate_path_tcx(tcx: TyCtxt, cnum: ast::CrateNum) -> PathBuf {
+    crate_path(tcx.sess, &tcx.crate_name(cnum), &tcx.crate_disambiguator(cnum))
+}
+
+/// Finds the correct metadata hashes file for the given crate. In order to do
+/// that it has to find the crate directory of the given crate, and in there,
+/// look for the session directory with the correct SVH in it.
```

Comment:
Do we not want some sort of read file-lock here? I worry that we will be in the middle of reading from a version of a crate when it is being recompiled. Are we not trying to defend against that scenario? Seems like it'd cause problems in other ways anyhow.